digdag-core/src/main/java/io/digdag/core/agent/RequireOperatorFactory.java

```diff
@@ -64,46 +68,109 @@ private RequireOperator(OperatorContext context, TaskCallbackApi callback)
+             *  If something errors happen, it will throw following exceptions.
+             *    - ResourceNotFoundException ... workflow ,project name(or id) are wrong. -> processed as deterministic error
+             *    - ResourceLimitExceededException ... this exception should be deterministic error
+             *    - SessionAttemptConflictException ... this is not error.
+             *      - If the conflict attempt is still running, wait for until done by nextPolling.
+             *      - If done, check the state param "require_kicked" and whether the attempt is kicked by this require> or not.
+             *        - If not kicked by this require>, check result and rerun_on option and determine rerun or not.
+             *          - if need to rerun, generate unique retry_attempt_name and set to "rerun_on_retry_attempt_name"
+             *          - throw nextPolling and in next call of runTask(), "rerun_on_retry_attempt_name" is used as retry_attempt_name and must succeed to create new attempt because it is unique.
+             *        - For both kicked and not kicked, check the result and ignore_failure option
+             *          - If ignore_failure is true or attempt finished successfully, require> op finished successfully
+             *          - else finished with exception.
+             */
             try {
-                ProjectIdentifier projectIdentifier = makeProjectIdentifier();
+                projectIdentifier = Optional.of(makeProjectIdentifier());
 
-                StoredSessionAttempt attempt = callback.startSession(
+                attempt = Optional.of(callback.startSession(
                         context,
                         request.getSiteId(),
-                        projectIdentifier,
+                        projectIdentifier.get(),
                         workflowName,
                         instant,
                         retryAttemptName,
-                        overrideParams);
+                        overrideParams));
+            }
+            catch (SessionAttemptConflictException ex) {
+                sessionAttemptConflictException = Optional.of(ex);
+            }
+            catch (ResourceNotFoundException ex) {
+                throw new TaskExecutionException(String.format(ENGLISH, "Dependent workflow does not exist. %s, workflowName:%s",
+                        projectIdentifier.transform((p)->p.toString()).or(""), workflowName ));
+            }
+            catch (ResourceLimitExceededException ex) {
+                throw new TaskExecutionException(ex);
+            }
 
-                boolean isDone = attempt.getStateFlags().isDone();
-                if (isDone) {
-                    if (!ignoreFailure && !attempt.getStateFlags().isSuccess()) {
+            if (sessionAttemptConflictException.isPresent()) {
+                StoredSessionAttempt conflictedAttempt = sessionAttemptConflictException.get().getConflictedSession();
```

Comment:
How about moving this block to `catch (SessionAttemptConflictException ex) { ... }` block? Maybe we can remove the Optional?

Reply:
Of course, we can do.  My intention of this way is that firstly get result of startSession() including exception (especially SessionAttemptConflictException is not error), then the logic goes ahead based on the result.
if we move the logic in `catch (SessionAttemptConflictException ex) { ... }`, catch {} become large.  
In this case it is better to define another method and call it.
What is your thought?

Reply:
Based on discussion in person, I fixed it.
As result, the code became simple. Thank you for your advice! 🙂